### PR TITLE
HPCC-26763 Add support for ROXIE TLS to kubernetes HPCC

### PR DESCRIPTION
**Pull request by afishbeck**
Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: esp/services/ws_ecl/ws_ecl_service.cpp [hunk 1]
```diff
@@ -196,22 +196,33 @@ static void appendServerAddress(StringBuffer &s, IPropertyTree &env, IPropertyTr
...
 
-        if (isEmptyString(target) || isEmptyString(name)) //bad config?
+        const char *target = service.queryProp("@target");
+        if (isEmptyString(target) || isEmptyString(service.queryProp("@name"))) //bad config?
             continue;
 
-        StringBuffer s;
-        s.append(name).append(':').append(port ? port : "9876");
-        Owned<ISmartSocketFactory> sf = new RoxieSocketFactory(s.str(), false, true, nullptr, (unsigned) -1);
+        Owned<ISmartSocketFactory> sf = new WsEclSocketFactory(service, "9876", false, true, nullptr, (unsigned) -1);
```
Comment (richardkchapman) [inline esp/services/ws_ecl/ws_ecl_service.cpp:225]: Should at least be using a constant from ports.h here - but not sure why you need a default port at all

File: esp/services/ws_ecl/ws_ecl_service.cpp [hunk 2]
```diff
@@ -2051,9 +2062,9 @@ void CWsEclBinding::sendRoxieRequest(const char *target, StringBuffer &req, Stri
         ep = conn->nextEndpoint();
 
         Owned<IHttpClientContext> httpctx = getHttpClientContext();
-        StringBuffer url("http://");
+        WsEclSocketFactory *roxieConn = static_cast<WsEclSocketFactory*>(conn);
+        StringBuffer url(roxieConn->isTlsService() ? "https://" : "http://");
         ep.getIpText(url).append(':').append(ep.port ? ep.port : 9876).append('/');
```
Comment (richardkchapman) [inline esp/services/ws_ecl/ws_ecl_service.cpp:2067]: Not new but same comment re port constant

File: esp/services/ws_smc/ws_smcService.cpp [hunk 3]
```diff
@@ -2229,7 +2229,7 @@ bool CWsSMCEx::onRoxieControlCmd(IEspContext &context, IEspRoxieControlCmdReques
     if (!isActiveK8sService(target))
         throw makeStringExceptionV(ECLWATCH_CANNOT_GET_ENV_INFO, "roxie target cluster has no active servers: %s", target);
 
-    Owned<IPropertyTree> controlResp = sendRoxieControlAllNodes(conn->nextEndpoint(), controlReq, true, req.getWait());
+    Owned<IPropertyTree> controlResp = sendRoxieControlAllNodes(conn, controlReq, true, req.getWait(), 1000);
```
Comment (richardkchapman) [inline esp/services/ws_smc/ws_smcService.cpp:2232]: What is 1000 and why is it hard-wired?

File: esp/smc/SMCLib/TpContainer.cpp [hunk 4]
```diff
@@ -689,16 +690,13 @@ extern TPWRAPPER_API void initContainerRoxieTargets(MapStringToMyClass<ISmartSoc
...
 
-        if (isEmptyString(target) || isEmptyString(name)) //bad config?
+        if (isEmptyString(target) || isEmptyString(service.queryProp("@name"))) //bad config?
             continue;
 
-        StringBuffer s;
-        s.append(name).append(':').append(port ? port : "9876");
-        Owned<ISmartSocketFactory> sf = new CSmartSocketFactory(s.str(), false, 60, (unsigned) -1);
+        bool tls = service.getPropBool("@tls", false);
+        Owned<ISmartSocketFactory> sf = tls ? createSecureSmartSocketFactory(service, "9876") : createSmartSocketFactory(service, "9876");
```
Comment (richardkchapman) [inline esp/smc/SMCLib/TpContainer.cpp:699]: Constants...

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 5]
```diff
@@ -2069,16 +2070,13 @@ extern TPWRAPPER_API void initContainerRoxieTargets(MapStringToMyClass<ISmartSoc
...
 
-        if (isEmptyString(target) || isEmptyString(name)) //bad config?
+        if (isEmptyString(target) || isEmptyString(service.queryProp("@name"))) //bad config?
             continue;
 
-        StringBuffer s;
-        s.append(name).append(':').append(port ? port : "9876");
-        Owned<ISmartSocketFactory> sf = new CSmartSocketFactory(s.str(), false, 60, (unsigned) -1);
+        bool tls = service.getPropBool("@tls", false);
+        Owned<ISmartSocketFactory> sf = tls ? createSecureSmartSocketFactory(service, "9876") : createSmartSocketFactory(service, "9876");
```
Comment (richardkchapman) [inline esp/smc/SMCLib/TpWrapper.cpp:2079]: constants...

File: helm/hpcc/templates/_helpers.tpl [hunk 6]
```diff
@@ -1535,7 +1568,7 @@ Pass in value
  {{- $number := (substr 0 (int (sub (len .) 1)) .) -}}
  {{- printf "%d" (int $number) -}}
 {{- else -}}
- {{- printf "%d" (int (mulf (float64 .) 1000.0)) -}}
+ {{- printf "%d" (int (1000000)) -}}
```
Comment (richardkchapman) [inline helm/hpcc/templates/_helpers.tpl:1571]: This change looks unrelated (and wrong!)

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Various port constants should be addressed

**Comment by richardkchapman** [comment]
There's a lot of changes here - worth getting a wider review I think

**Comment by richardkchapman** [comment]
Also probably beyond the scope of a point release - needs to target master

File: esp/clients/roxiecontrol.cpp [hunk 7]
```diff
@@ -77,6 +79,46 @@ IPropertyTree *sendRoxieControlQuery(ISocket *sock, const char *msg, unsigned wa
     return ret.getClear();
 }
 
+IPropertyTree *sendRoxieControlQuery(const SocketEndpoint &ep, const char *msg, unsigned wait, unsigned connect_wait)
+{
+    Owned<ISocket> sock = ISocket::connect_timeout(ep, connect_wait);
+    return sendRoxieControlQuery(sock, msg, wait);
+}
+
+IPropertyTree *sendSecureRoxieControlQuery(ISmartSocketFactory *conn, const char *msg, unsigned wait, unsigned connect_wait)
```
Comment (ghalliday) [inline esp/clients/roxiecontrol.cpp:88]: Does this need a #ifdef _USE_OPENSSL?

File: esp/clients/roxiecontrol.cpp [hunk 8]
```diff
@@ -111,3 +153,37 @@ IPropertyTree *sendRoxieControlAllNodes(const SocketEndpoint &ep, const char *ms
     Owned<ISocket> sock = ISocket::connect_timeout(ep, wait);
     return sendRoxieControlAllNodes(sock, msg, allOrNothing, wait);
 }
+
+IPropertyTree *sendSecureRoxieControlAllNodes(ISmartSocketFactory *conn, const char *msg, bool allOrNothing, unsigned wait, unsigned connect_wait, bool publicService)
+{
+    const SocketEndpoint &ep = conn->nextEndpoint();
```
Comment (ghalliday) [inline esp/clients/roxiecontrol.cpp:159]: most f this code looks identical to the code in sendRoxieControlQuery(), and a candidate for being shared.

Comment (ghalliday) [reply esp/clients/roxiecontrol.cpp:159]: if conn->isTlsService() was checked within this function it would probably simplify the calling logic.

File: esp/services/ws_dfu/ws_dfuXRefService.cpp [hunk 9]
```diff
@@ -715,8 +715,7 @@ bool CWsDfuXRefEx::onDFUXRefUnusedFiles(IEspContext &context, IEspDFUXRefUnusedF
     if (!servers.length())
         throw MakeStringExceptionDirect(ECLWATCH_INVALID_CLUSTER_INFO, "process cluster, not found.");
 
-    Owned<ISocket> sock = ISocket::connect_timeout(servers.item(0), 5000);
-    Owned<IPropertyTree> controlXrefInfo = sendRoxieControlQuery(sock, "<control:getQueryXrefInfo/>", 5000);
+    Owned<IPropertyTree> controlXrefInfo = sendRoxieControlQuery(servers.item(0), "<control:getQueryXrefInfo/>", 5000, 1000);
```
Comment (ghalliday) [inline esp/services/ws_dfu/ws_dfuXRefService.cpp:718]: Would a simpler change have been 

`Owned<ISocket> sock = createRoxieControlSocket(servers.item(0), 5000);`

File: esp/services/ws_ecl/ws_ecl_service.cpp [hunk 10]
```diff
@@ -275,7 +286,7 @@ void initBareMetalRoxieTargets(MapStringToMyClass<ISmartSocketFactory> &connMap,
         if (list.length())
         {
             StringAttr alias(clusterInfo->getAlias());
-            Owned<ISmartSocketFactory> sf = new RoxieSocketFactory(list.str(), !loadBalanced, includeTargetInURL, loadBalanced ? alias.str() : NULL, dnsInterval);
+            Owned<ISmartSocketFactory> sf = (ISmartSocketFactory *) new WsEclSocketFactory(list.str(), !loadBalanced, includeTargetInURL, loadBalanced ? alias.str() : NULL, dnsInterval);
```
Comment (ghalliday) [inline esp/services/ws_ecl/ws_ecl_service.cpp:289]: minor: is the cast needed?

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 11]
```diff
@@ -1274,14 +1274,17 @@ IPropertyTree *getQueriesOnCluster(const char *target, const char *queryset, Str
             control.append("</control:queries>");
         }
 #ifndef _CONTAINERIZED
-        Owned<ISocket> sock = ISocket::connect_timeout(eps.item(0), ROXIECONNECTIONTIMEOUT);
+        const SocketEndpoint &ep = eps.item(0);
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp:1277]: minor: ep not used (probably meant to in 1279/1281)

File: roxie/ccd/ccdmain.cpp [hunk 12]
```diff
@@ -1355,28 +1359,39 @@ int CCD_API roxie_main(int argc, const char *argv[], const char * defaultYaml)
                     if (!roxiePort)
                     {
                         roxiePort = port;
+                        if (roxieFarm.getPropBool("@tls"))
+                            roxiePortTlsClientConfig = createTlsClientSecretInfo(roxieFarm.queryProp("@issuer"), roxieFarm.getPropBool("@pulic")==false, roxieFarm.getPropBool("@selfSigned"));
```
Comment (ghalliday) [inline roxie/ccd/ccdmain.cpp:1363]: typo: "@pulic".
!x.getPropBool() is also more common/clearer than x.getPropBool()==false?

File: system/jlib/jsecrets.cpp [hunk 13]
```diff
@@ -764,6 +764,51 @@ const MemoryAttr &getSecretUdpKey(bool required)
...
+            return nullptr;
+
+        info->setProp("certificate", filepath.str());
+        filepath.set(secretpath).append("tls.key");
+        if (checkFileExists(filepath))
+            info->setProp("privatekey", filepath.str());
+    }
+
+    IPropertyTree *verify = ensurePTree(info, "verify");
+    if (verify)
```
Comment (ghalliday) [inline system/jlib/jsecrets.cpp:792]: can this ever be nulll?

File: system/jlib/jsecrets.cpp [hunk 14]
```diff
@@ -764,6 +764,51 @@ const MemoryAttr &getSecretUdpKey(bool required)
...
+    {
+        if (addCACert)
+        {
+            filepath.set(secretpath).append("ca.crt");
+            if (checkFileExists(filepath))
+            {
+                IPropertyTree *ca = ensurePTree(verify, "ca_certificates");
+                if (ca)
```
Comment (ghalliday) [inline system/jlib/jsecrets.cpp:800]: same here, not sure this test is needed

File: system/jlib/jsmartsock.cpp [hunk 15]
```diff
@@ -449,6 +480,15 @@ StringBuffer & CSmartSocketFactory::getUrlStr(StringBuffer &url, bool useHostNam
...
+ISmartSocketFactory *createSmartSocketFactory(IPropertyTree &service, const char *defPort, bool _retry, unsigned _retryInterval, unsigned _dnsInterval)
+{
+    const char *name = service.queryProp("@name");
+    DBGLOG("creating smart socket service client: %s, %s", service.getPropBool("@tls") ? "manual TLS" : "TCP", name ? name : "unnamed");
+    return new CSmartSocketFactory(service, defPort, _retry, _retryInterval, _dnsInterval);
+}
+
+ISmartSocketFactory *createSmartSocketFactory(const char *_socklist, bool _retry, unsigned _retryInterval, unsigned _dnsInterval)
+{
+    DBGLOG("creating smart socket: %s", _socklist);
```
Comment (ghalliday) [inline system/jlib/jsmartsock.cpp:492]: Could this tracing become excessive?  Same for the function above.  Possibly not since it is a factory

File: system/jlib/jsmartsock.ipp [hunk 16]
```diff
@@ -97,6 +101,8 @@ public:
     virtual void resolveHostnames();
 
     virtual StringBuffer & getUrlStr(StringBuffer &str, bool useHostName);
+    virtual bool isTlsService() const override {return tlsService;}
```
Comment (ghalliday) [inline system/jlib/jsmartsock.ipp:104]: trivial: Normally have whitespace after { and before }

File: system/security/securesocket/securesocket.cpp [hunk 17]
```diff
@@ -1999,9 +2020,16 @@ class CSecureSmartSocketFactory : public CSmartSocketFactory
 
 ISmartSocketFactory *createSecureSmartSocketFactory(const char *_socklist, bool _retry, unsigned _retryInterval, unsigned _dnsInterval)
 {
+    DBGLOG("createing smart socket: TLS, %s", _socklist);
     return new CSecureSmartSocketFactory(_socklist, _retry, _retryInterval, _dnsInterval);
 }
 
+ISmartSocketFactory *createSecureSmartSocketFactory(IPropertyTree &service, const char *defPort, bool _retry, unsigned _retryInterval, unsigned _dnsInterval)
+{
+    DBGLOG("createing service smart socket: TLS");
```
Comment (ghalliday) [inline system/security/securesocket/securesocket.cpp:2029]: Could this tracing become excessive?  Same for the function above.  (Also typo in creat**e**ing )

**Review by ghalliday — COMMENTED** [review]
@afishbeck some other comments - mainly #ifdef open_ssh and commoning up the code.

**Comment by afishbeck** [comment]
@ghalliday @richardkchapman updated based on comments.  Also refactored the helm chart templates.  I was afraid it was too dependent on order the other way.

**Comment by afishbeck** [comment]
Retargeted to master.

File: esp/clients/roxiecontrol.hpp [hunk 18]
```diff
@@ -21,11 +21,19 @@
 #include "jlib.hpp"
 #include "jsocket.hpp"
 #include "jptree.hpp"
-#include "roxiecontrol.hpp"
+#include "jsmartsock.hpp"
+
+const unsigned ROXIECONNECTIONTIMEOUT = 1000;   //1 second
+const unsigned ROXIECONTROLQUERYTIMEOUT = 3000; //3 second
```
Comment (ghalliday) [inline esp/clients/roxiecontrol.hpp:27]: What is the difference between this and the next timeout?

File: helm/hpcc/templates/roxie.yaml [hunk 19]
```diff
@@ -33,7 +34,18 @@ data:
...
     version: 1.0
     roxie:
-{{ toYaml ( omit .me "logging" "topoServer" "encryptInTransit" "env") | indent 6 }}
+{{- $root := .root -}}
+{{- $component := .me }}
+      services:
+  {{- range $service := .me.services }}
+      - name: {{ $service.name }}
+{{ toYaml (omit $service "tls" "name") | indent 8 }}
+
```
Comment (ghalliday) [inline helm/hpcc/templates/roxie.yaml:43]: trivial: I suspect this newline is not deliberate

File: system/jlib/jsmartsock.cpp [hunk 20]
```diff
@@ -207,6 +208,39 @@ void CSmartSocket::close()
...
+    tlsService  = service.getPropBool("@tls");
+    if (tlsService)
+        tlsConfig.setown(createTlsClientSecretInfo(service.queryProp("@issuer"), service.getPropBool("@pulic")==false, service.getPropBool("@selfSigned"), service.getPropBool("@caCert")));
+
+    StringBuffer s;
+    s.append(name).append(':').append(port);
+
+    PROGLOG("CSmartSocketFactory::CSmartSocketFactory(service(%s), %s)", name, s.str());
+
+
```
Comment (ghalliday) [inline system/jlib/jsmartsock.cpp:227]: trivial: extra newline

**Review by ghalliday — APPROVED** [review]
Looks good - only a few trivial comments

File: esp/clients/roxiecontrol.hpp (hunk 18 above)
Comment (afishbeck) [reply esp/clients/roxiecontrol.hpp:27]: I may have misinterpreted it slightly.  These were existing values that I moved, and I thought they were general purpose, (just badly named), but the `ROXIECONTROLQUERIESTIMEOUT ` one at least looks more specific to `"<control:queries>"`.  I'll add new ones where appropriate to avoid confusion.

File: esp/services/ws_ecl/ws_ecl_service.cpp (hunk 2 above)
Comment (afishbeck) [reply esp/services/ws_ecl/ws_ecl_service.cpp:2067]: I changed the others but this one could affect baremetal, so I'd rather leave it, unless you want me to investigate further.

File: esp/clients/roxiecontrol.hpp (hunk 18 above)
Comment (afishbeck) [reply esp/clients/roxiecontrol.hpp:27]: Adding a value specific to `control:getQueryXrefInfo` and assigning the value 5000 which is what's been used for that query in the past.

**Comment by afishbeck** [comment]
Applied a couple of minor changes.

File: system/jlib/jsecrets.cpp [hunk 21]
```diff
@@ -764,6 +764,48 @@ const MemoryAttr &getSecretUdpKey(bool required)
...
+
+    StringBuffer filepath;
+    StringBuffer secretpath;
+    buildSecretPath(secretpath, "certificates", issuer);
+
+    Owned<IPropertyTree> info = createPTree();
+
+    if (mutual)
+    {
+
```
Comment (jakesmith) [inline system/jlib/jsecrets.cpp:780]: trivial: unecessary blank line

File: system/jlib/jsmartsock.cpp [hunk 22]
```diff
@@ -207,6 +208,38 @@ void CSmartSocket::close()
...
+CSmartSocketFactory::CSmartSocketFactory(IPropertyTree &service, bool _retry, unsigned _retryInterval, unsigned _dnsInterval)
+{
+    const char *name = service.queryProp("@name");
+    const char *port = service.queryProp("@port");
+    if (isEmptyString(name) || isEmptyString(port))
+        throw createSmartSocketException(0, "CSmartSocket factory both name and port required for service configuration");
+
+    tlsService  = service.getPropBool("@tls");
+    if (tlsService)
+        tlsConfig.setown(createTlsClientSecretInfo(service.queryProp("@issuer"), service.getPropBool("@pulic")==false, service.getPropBool("@selfSigned"), service.getPropBool("@caCert")));
```
Comment (jakesmith) [inline system/jlib/jsmartsock.cpp:220]: trivial (and prev. comment from Gavin): more standard/clearer to use !service.getPropBool("@public")

**Review by jakesmith — COMMENTED** [review]
have scanned through, bar 2 trivial comments, I can't see any issues

**Review by richardkchapman — APPROVED** [review]

**Review by mckellyln — APPROVED** [review]

**Comment by richardkchapman** [comment]
@afishbeck Needs rebasing for a clean merge

**Comment by afishbeck** [comment]
@richardkchapman rebased and fixed conflicts.
